src/frontend/src/optimizer/optimizer_context.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::convert::{From, Into};
+use core::fmt::Formatter;
+use core::marker::Sync;
+use core::sync::atomic::{AtomicBool, AtomicI32, AtomicU32, Ordering};
+use std::cell::RefCell;
+use std::sync::Arc;
+
+use crate::expr::CorrelatedId;
+use crate::handler::HandlerArgs;
+use crate::optimizer::plan_node::PlanNodeId;
+use crate::session::SessionImpl;
+use crate::WithOptions;
+
+pub struct OptimizerContext {
+    pub session_ctx: Arc<SessionImpl>,
+    // We use `AtomicI32` here because `Arc<T>` implements `Send` only when `T: Send + Sync`.
+    pub next_id: AtomicI32,
+    /// For debugging purposes, store the SQL string in Context
+    pub sql: Arc<str>,
+
+    /// it indicates whether the explain mode is verbose for explain statement
+    pub explain_verbose: AtomicBool,
+
+    /// it indicates whether the explain mode is trace for explain statement
+    pub explain_trace: AtomicBool,
+    /// Store the trace of optimizer
+    pub optimizer_trace: RefCell<Vec<String>>,
```

Comment:
I guess we're able to replace all `Atomic` to `RefCell` and `Arc` to `Rc` as well now, or even directly wrap the `OptimizerContext` with `RefCell`?

Reply:
Yes, I think we can use `RefCell` to replace `Atomic ` and `Arc` to `Rc`,  except `Arc<SessionImpl>`. Wrapping the `OptimizerContext ` with `RefCell` sounds attractive.

Reply:
After second thought, wrapping the `OptimizerContext` with `RefCell` directly seems a little bit dangerous, because it could panic if the outside code uses it casually. I think it is better to wrap the member with `RefCell` if we need to modify it and provide some methods instead of letting users access the `RefCell` directly.